keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1210,230 +1099,251 @@ server-side-apply (SSA) simpler. Patching individual list elements is not
-type ResourceModel struct {
-    // NamedResources describes available resources using the named resources model.
-    NamedResources *NamedResourcesResources
-}
-```
-
-The "named resources" model lists individual resource instances and their
-attributes:
-
 ```go
-// NamedResourcesResources is used in NodeResourceModel.
-type NamedResourcesResources struct {
-    // The list of all individual resources instances currently available.
-    Instances []NamedResourcesInstance
-}
-
-// NamedResourcesInstance represents one individual hardware instance that can be selected based
+// Device represents one individual hardware instance that can be selected based
 // on its attributes.
-type NamedResourcesInstance struct {
-    // Name is unique identifier among all resource instances managed by
-    // the driver on the node. It must be a DNS subdomain.
+type Device struct {
+    // Name is unique identifier among all devices managed by
+    // the driver in the pool. It must be a DNS label.
     Name string
 
-    // Attributes defines the attributes of this resource instance.
+    // Attributes defines the set of attributes for this device.
     // The name of each attribute must be unique.
-    Attributes []NamedResourcesAttribute
+    //
+    // Must not have more than 32 entries.
+    //
+    // +listType=atomic
+    // +optional
+    Attributes []DeviceAttribute
+
+    // SharedCapacityConsumed defines the set of shared capacity consumed by
+    // this device.
+    //
+    // Must not have more than 32 entries.
+    //
+    // +listType=atomic
+    // +optional
+    SharedCapacityConsumed []SharedCapacity
 }
 
-// NamedResourcesAttribute is a combination of an attribute name and its value.
-type NamedResourcesAttribute struct {
-    // Name is unique identifier among all resource instances managed by
-    // the driver on the node. It must be a DNS subdomain.
+const ResourceSliceMaxAttributesPerDevice = 32
+const ResourceSliceMaxSharedCapacityConsumedPerDevice = 32
+
+// ResourceSliceMaxDevices and ResourceSliceMaxAttributesPerDevice were chosen
+// so that with a maximum `DeviceAttribute` length of 96 characters and a
+// maximum `SharedCapacity` length of ~40 characters (ignoring overhead), the
+// total size of the ResourceSlice object is around 590KB.
+
+// DeviceAttribute is a combination of an attribute name and its value.
+// Exactly one value must be set.
+type DeviceAttribute struct {
+    // Name is a unique identifier for this attribute, which will be
+    // referenced when selecting devices.
+    //
+    // Attributes are defined either by the owner of the specific driver
+    // (usually the vendor) or by some 3rd party (e.g. the Kubernetes
+    // project). Because attributes are sometimes compared across devices,
+    // a given name is expected to mean the same thing and have the same
+    // type on all devices.
+    //
+    // Attribute names must be either a C-style identifier
```

Comment:
I agree. It is much nicer to not be limited to C style identifiers here.

Reply:
+1

Reply:
Agreed. I hadn't dropped it yet because the decision about CEL was still a bit vague, but as we now all agree I'll convert to DNS label. I will keep a max length of 32 characters, though, because our worst-case analysis of object size is very sensitive to attribute name length.